### ISSUE TITLE
AI employ diplomats and take their success into account

### ISSUE THREAD

**Issue by LynxAbraxas** [civctp2/civctp2#112]
So far I never observed that the AI tried to establish an embassy or hold a reception.
Due to this the AI does not know the scientific advances of other players and therefore cannot suggest diplomatic exchange of these.

**Comment by MartinGuehmann** [comment]
So at least, the AI now builds diplomats with 31f0ae6. I still haven't seen the AI using them for holding receptions (throwing parties) or establishing embassies.

**Comment by LynxAbraxas** [comment]
Really cool! I'm wondering though what "missionaries" are. I only know clerics and televangelists or is "missionaries" used for both, since televangelists replace clerics when advancing? I'd love to see some of the AI civs to stay in theology longer ideally till it can build televangelists. They just such a cool game concept special to CTP.

**Comment by MartinGuehmann** [comment]
> Really cool! I'm wondering though what "missionaries" are. I only know clerics and televangelists or is "missionaries" used for both, since televangelists replace clerics when advancing? I'd love to see some of the AI civs to stay in theology longer ideally till it can build televangelists. They just such a cool game concept special to CTP.

I used missionary as a category, to put in those people. A televangelist, evangelizes or proselytizes, basically that is also what the cleric does here CTP2, even so a cleric is not necessarily somebody who evangelize. Simply these people spread the word of their religion, whatever that is.

If you wanna see the televangelists in the game then properly we should allow them either for all governments or at least put the televangelists into more modern governments.

**Comment by LynxAbraxas** [comment]
> put the televangelists into more modern governments.

Hm, is there a modern form of theology? If not, how much work would it be to create one (that would possibly allow more cities).
Putting it in all govs probably renders theology uninteresting.

**Comment by MartinGuehmann** [comment]
> > put the televangelists into more modern governments.
> 
> Hm, is there a modern form of theology? If not, how much work would it be to create one (that would possibly allow more cities).

There are modern theocracies such as Saudi Arabia and the Iran. But well, televangelists are not restricted to theocracies, you also find them in the USA, they are not really known here in Germany.

> Putting it in all govs probably renders theology uninteresting.

Theocracy is for me rather uninteresting, anyway. But of course that could be changed, by modifying it in other ways.

**Comment by LynxAbraxas** [comment]
Sorry, I meant a modern form of Theocracy in ctp2. For me theocracy in ctp2 is interesting because of clerics and later televangelists which allow a form of game interaction between the extremes of military, science and economy. Basically, they combine unit action without destruction, and their conversion animation is really cool (in contrast to e.g. the corporate branch).
Sticking to ctp2 concepts, how about allowing other govs to build missionaries if they have built a specific wonder? 
Or to make them accessible to multiple players in the same game one could think of a minimum number of e.g. basilicas for clerics  and a min. number of e.g. TVs for televangelists?

**Comment by LynxAbraxas** [comment]
Seems we got off topic. Missionaries in other govs than theocracy should be in its own issue: #129 
> I still haven't seen the AI using them for holding receptions (throwing parties) or establishing embassies.

It' great to have the AI build them with https://github.com/civctp2/civctp2/commit/31f0ae688b5a0674a8f04988bed0db2c63022324, still I would say the issue is not closed as long as the AI also uses their special abilites. Receptions probably being the easier one but also the less interesting. In the end it would be great if the AI would use its diplomats to establish embassies AND to make use of the newly gained info in its behaviour, which is probably much more difficult to implement (unless the code is already ther but not used so far).

**Comment by MartinGuehmann** [comment]
> Seems we got off topic. Missionaries in other govs than theocracy should be in its own issue: #129
> 
> > I still haven't seen the AI using them for holding receptions (throwing parties) or establishing embassies.
> 
> It' great to have the AI build them with [31f0ae6](https://github.com/civctp2/civctp2/commit/31f0ae688b5a0674a8f04988bed0db2c63022324), still I would say the issue is not closed as long as the AI also uses their special abilites. Receptions probably being the easier one but also the less interesting. In the end it would be great if the AI would use its diplomats to establish embassies AND to make use of the newly gained info in its behaviour, which is probably much more difficult to implement (unless the code is already ther but not used so far).

Well, I am sure that the code is already there. It just does not do what it is supposed to do.

For the receptions, the scheduler didn't add any units, but it added all the cities of the other AIs even those my AI was at war with. So there is definitively something wrong. However, receptions improve the regard of the other AI players among each other and thus they would be more likely to be at piece and thus would be able to establish embassies.

**Comment by MartinGuehmann** [comment]
Well, actually they are throwing parties and establish embassies. I just don't get, why they are even doing it with civs you they are at war with.

**Comment by LynxAbraxas** [comment]
> Well, actually they are throwing parties and establish embassies.

That's really cool. So you already observe that with just https://github.com/civctp2/civctp2/commit/31f0ae688b5a0674a8f04988bed0db2c63022324 or did you test with additional changes?

> I just don't get, why they are even doing it with civs you they are at war with.

Hm, are they just attempting or are they actually succeeding with that? If they just attempt and then the action is blocked by the rule that it is not possible while at war, this could be bearable. In that case it would "just" be extra CPU load but no "cheating".

**Comment by MartinGuehmann** [comment]
> > Well, actually they are throwing parties and establish embassies.
> 
> That's really cool. So you already observe that with just [31f0ae6](https://github.com/civctp2/civctp2/commit/31f0ae688b5a0674a8f04988bed0db2c63022324) or did you test with additional changes?

I modified the goal entry, by removing the stealth class requirement, since diplomats aren't really stealth units. However, I saw in the save game that another AI had two embassies with two small civs, and since I just had reloaded the game they couldn't have established them, since the change.

My main AI did not seek to establish embassies, because it has the United Nations, so it has embassies, anyway. The thing with the receptions is that I am unfamiliar with the special effect, so I could even have it seen more often and you cannot hold receptions every turn, you have to wait some turn when you can give receptions again.

> > I just don't get, why they are even doing it with civs you they are at war with.
> 
> Hm, are they just attempting or are they actually succeeding with that? If they just attempt and then the action is blocked by the rule that it is not possible while at war, this could be bearable. In that case it would "just" be extra CPU load but no "cheating".

They didn't succeed. However, it is still a waste of resources, because instead of this you could go to another civ with that you have peace, in fact these two goals target everybody except hot war enemies. So diplomats shouldn't go to cities of civs with them you have war.

**Comment by MartinGuehmann** [comment]
It is indeed the problem that diplomats do not belong to squad class stealth. So removing that from the goal fixes this, we still have the other problems.

**Comment by MartinGuehmann** [comment]
7eb8453 allows the AI to throw parties and to establish embassies.

9f938c1 keeps the AI from trying this with civs it is at war with.

**Comment by LynxAbraxas** [comment]
Wonderful, I'll test them. Looking forward to see some diplomats form the AI.
Why are diplomats no stealth units in ctp2? Because of this it can be a real hassle protecting them to get them do their job.

**Comment by MartinGuehmann** [comment]
> Why are diplomats no stealth units in ctp2? Because of this it can be a real hassle protecting them to get them do their job.

I think that because diplomacy is considered a legal activity, while planting nukes, stealing slaves, etc. are things thou shalt not do.

But of course then the AI shouldn't kick them out, either.

**Comment by MartinGuehmann** [comment]
Should then the AI ignore diplomats and not kick out if it is not at war with their civ?

**Comment by LynxAbraxas** [comment]
> Should then the AI ignore diplomats and not kick out if it is not at war with their civ?

Sounds like a good idea. Otherwise you have to acompany them with a military unit for protection which then is not good for the diplomatic relationship.

**Comment by LynxAbraxas** [comment]
All obvious stuff seems to be solved with #119 (and #177). The AI installes embassies and holds parties:

![ss_2019-09-29_23:24:25](https://user-images.githubusercontent.com/21012234/66706978-5915d580-ed3a-11e9-8b34-21612853bb60.png)

But does the AI take held parties and info gained through embassies into account?
@MartinGuehmann any idea if that is in the code anywhere?

**Comment by MartinGuehmann** [comment]
> All obvious stuff seems to be solved with #119 (and #177). The AI installes embassies and holds parties:

> But does the AI take held parties and info gained through embassies into account? @MartinGuehmann any idea if that is in the code anywhere?

Holding parties just improves the regard, which is of course taken into account and the information gained by embassies is known by the AI anyway. And thus taken into account anyway.

Looks like we can close this issue.
